### ISSUE TITLE
Exclude files in tsconfig without throwing an @typescript-eslint/parser error

### ISSUE THREAD

**Issue by leanne1** [typescript-eslint/typescript-eslint#1350]
I know this has been raised in various issues but I do not understand how to handle this error.
I am using typescript-eslint using type checking rules in the linter and have set my lint to use my tsconfig for type checks. The issue is I have a glob I want to exclude from TS checks, which I do in tsconfig and works correctly in ts compiler (as in these files are not type checked), but as soon as i exclude these files from TS this throws many errors in the linter e.g. 

```error  Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser```

I want to be able to exclude files from TS without this error getting thrown by the linter and *without disabling all other linting on the glob* which right now is the only solution i can think of to move forward (or removing lint type checking altogether).

Any advice on how to fix appreciated.

```ts
// tsconfig
{
  "compileOnSave": true,
  "compilerOptions": {
    "outDir": "./build",
    "allowJs": true,
    "checkJs": false,
    "esModuleInterop": false,
    "allowSyntheticDefaultImports": true,
    "target": "ES2018",
    "module": "commonjs",
    "strict": true,
    "noUnusedLocals": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "inlineSourceMap": true,
    "inlineSources": false,
    "experimentalDecorators": true,
    "jsx": "react",
    "importHelpers": true,
    "moduleResolution": "node",
    "skipLibCheck": true,
    "forceConsistentCasingInFileNames": true,
    "emitDecoratorMetadata": true,
    "noImplicitThis": false /*webdriverio configuration*/,
    "alwaysStrict": true /*webdriverio configuration*/,
    "removeComments": true,
    "baseUrl": ".",
    "paths": {
      "*": ["./*"],
      "steps/*": ["./test/*"]
    }
  },
  "include": ["./rtl.setup.js", "./src/**/*.ts", "./src/**/*.tsx", "./test/**/*.ts"],
  "exclude": ["node_modules", "build", "config", "doc", "env", "public", "scripts", "./src/**/*.test.ts"] // => ./src/**/*.test.ts" is the glob exclusion that is causing the lint errors
}

```

```js
// .eslintrc

const ERROR = 2;
const OFF = 0;

const MAX_PARAMS = 4;
const MAX_LINE_LENGTH = 120;
const MAX_NESTED_CALLBACKS = 3;
const MAX_STATEMENTS = 30;

module.exports = {
  parser: "@typescript-eslint/parser",
  parserOptions: {
    ecmaVersion: 2020,
    sourceType: "module",
    ecmaFeatures: {
      jsx: true,
    },
    project: "./tsconfig.json",
    tsconfigRootDir: __dirname,
  },
  plugins: ["@typescript-eslint", "react", "jsx-a11y", "prettier", "promise"],
  extends: [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:@typescript-eslint/recommended-requiring-type-checking",
    "airbnb",
    "airbnb/hooks",
    "plugin:jsx-a11y/recommended",
    "plugin:import/typescript",
    "plugin:node/recommended",
    "prettier",
    "prettier/@typescript-eslint",
    "plugin:promise/recommended",
  ],
  env: {
    browser: true,
    node: true,
    jest: true,
    "jest/globals": true,
  },
  settings: {
    react: {
      version: "detect",
    },
    "import/resolver": {
      node: {
        extensions: [".js", ".jsx", ".ts", ".tsx"],
      },
    },
  },
  globals: {
    __APOLLO_STATE__: "readonly",
    WEBPACK_ENV: "readonly",
  },
  rules: {
    "max-len": [ERROR, MAX_LINE_LENGTH],
    "max-params": [ERROR, MAX_PARAMS],
    "no-nested-ternary": OFF,
    "no-underscore-dangle": OFF, // Used in external APIs
    camelcase: OFF, // It doesn't work well with typescript interfaces for external APIs
    quotes: [ERROR, "double", { avoidEscape: true }],
    "no-return-assign": [ERROR, "always"],
    "class-methods-use-this": OFF, // We should not check for "this" in public methods.
    "quote-props": OFF,
    "array-element-newline": [ERROR, "consistent"],
    "comma-dangle": [ERROR, "always-multiline"],
    "function-paren-newline": [ERROR, "multiline-arguments"],
    "max-nested-callbacks": [ERROR, MAX_NESTED_CALLBACKS],
    "max-statements": [ERROR, MAX_STATEMENTS],
    "func-style": [ERROR, "declaration", { allowArrowFunctions: true }],
    "consistent-this": [ERROR, "self"],
    "no-cond-assign": [ERROR, "always"],
    "no-shadow": [
      ERROR,
      {
        hoist: "all",
      },
    ],
    "space-before-function-paren": [
      "error",
      {
        anonymous: "never",
        named: "never",
        asyncArrow: "always",
      },
    ],
    "padded-blocks": [ERROR, "never"],
    "no-empty-function": [ERROR, { allow: ["constructors"] }],
    "no-unused-vars": OFF, // Delegate to @typescript-eslint

    // Plugin rules
    "@typescript-eslint/no-unused-vars": ERROR,
    "react/jsx-filename-extension": [ERROR, { extensions: [".tsx"] }],
    "react/jsx-props-no-spreading": OFF,
    "react/jsx-one-expression-per-line": OFF, // clashes with prettier
    "react-hooks/rules-of-hooks": ERROR,
    "jsx-a11y/label-has-associated-control": [
      ERROR,
      {
        labelComponents: ["Label"],
        controlComponents: ["TextInput"],
      },
    ],
    "import/prefer-default-export": OFF,
    "import/no-default-export": ERROR,
    "import/no-extraneous-dependencies": [ERROR, { devDependencies: true }],
    "prettier/prettier": ERROR,
    "node/no-unsupported-features/es-syntax": OFF,
    "node/no-unsupported-features/es-builtins": ["error", { version: "10.x" }],
    "node/no-unsupported-features/node-builtins": ["error", { version: "10.x" }],
  },

  overrides: [
    // Javascript files
    {
      files: ["*.js"],
      rules: {
        "@typescript-eslint/explicit-function-return-type": OFF,
        "global-require": OFF,
      },
    },
    // Typescript files
    {
      files: ["*.ts", "*.tsx"],
      rules: {
        "@typescript-eslint/explicit-function-return-type": ERROR,
        "@typescript-eslint/no-explicit-any": ERROR,
        "@typescript-eslint/camelcase": OFF,
      },
    },
    // Unit test files
    {
      files: ["*.test.js", "*.test.ts", "*.test.tsx"],
      globals: {
        jsdom: "readonly",
      },
      plugins: ["jest"],
      extends: ["plugin:jest/recommended"],
      rules: {
        "jest/require-top-level-describe": ERROR,
        "jest/no-try-expect": OFF,
      },
    },
    // e2e test files
    {
      files: ["./test/**/*.ts"],
      globals: {
        browser: "readonly",
        $: "readonly",
      },
      plugins: ["chai-friendly"],
      rules: {
        "@typescript-eslint/no-var-requires": OFF,
        "func-names": OFF,
        "no-unused-expressions": OFF,
        "chai-friendly/no-unused-expressions": ERROR,
      },
    },
  ],
};

```

**What did you expect to happen?**
I want to be able to exclude files from TS checks without a lint error being thrown

**What actually happened?**
I get many lint errors, all relating to test files i.e.

```
/xx/xxx/xxx/xxx/xxx/src/server/schema/enum/PublicationType.test.ts
0:0  error  Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser.
The file does not match your project config: src/server/schema/enum/PublicationType.test.ts.
The file must be included in at least one of the projects provided
```

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `2.11.0` |
| `TypeScript`                | `3.6.2` |
| `ESLint`                    | `6.3.0` |
| `node`                      | `10.15.1` |
| `npm`                       | `6.13.1` |

**Comment by bradzacher** [comment]
Have a look at our onboarding docs.

https://github.com/typescript-eslint/typescript-eslint/tree/master/docs/getting-started/linting#ignoring-unnecessary-files

**Comment by leanne1** [comment]
Thanks @bradzacher for you quick reply. However if  i understand correctly you are suggesting I lint-ignore the whole glob? As mentioned initially this is what i want to avoid doing, I have other lint rules which target this file glob and I wish to keep using. My question is how to stop the linter throwing a parser error when i TS ignore some files in the project? This also feels like a bug to me as the error is for about 28 files but i have more than 28 *.test.ts files matching the exclude glob in my project, so why is the linter objecting to just some?

**Comment by bradzacher** [comment]
There are more docs in that folder which might help you. See the docs on typed linting.

It depends on your exact intentions.

If you want to exclude the files from your tsconfig but not from the lint, then create a new tsconfig which is only used for linting.

If you want to exclude the files from your tsconfig and the lint, then use an eslint ignore file.

**Comment by leanne1** [comment]
Thanks again for the additional information @bradzacher I fixed the issue by using a separate tsconfig just for the eslint config to use.

To be clear for others with the same issue as it wasn't clear to me from the existing issues i read, the following can be used to turn off lint file parse errors:

1. Create a separate tsconfig (e.g `tsconfig.eslint.json` for use only by the the lint `parserOptions.project` value in your `.eslintrc`
2. In `tsconfig.eslint.json` extend from your original `tsconfig.json`
3. Include both an `includes` and `excludes` value in `tsconfig.eslint.json`. `includes`  must include all the files in your project and `excludes` must include all the files in your original `tsconfig.json` `excludes` *minus the files you are excluding from TS checks`. I also found an empty array also worked.

Based on my config above, the following `tsconfig.eslint.json` fixed my issue:
```
{
  "extends": "./tsconfig.json",
  "include": ["rtl.setup.js", "./src/**/*.ts", "./src/**/*.tsx", "./test/**/*.ts"],
  "exclude": ["node_modules", "build", "config", "doc", "env", "public", "scripts"] // or: []
}
```

**Comment by bradzacher** [comment]
https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/MONOREPO.md#one-root-tsconfigjson

Sorry, it's in the monorepo docs, because it's rare that single project repos use or need this kind of more complex config.

>  I also found an empty array also worked.

I suggest not doing this. An empty `include` array is the same as providing `include: ["**/*"]`, which is more than likely going to include more files than you want, and slow down your lint runs.

---

I'll close this off then as you've resolved it.
